The Ember Times No. 162 - August 28th 2020 (#751)
ember-learn/ember-blog: source/2020-08-28-the-ember-times-issue-162.md

[Chris Manson (@mansona)](https://github.com/mansona) released the first working version of `ember-showdown-prism` this week. This drop-in addon automatically adds [Prism](https://prismjs.com/) syntax highlighting to Markdown code blocks if you are using [showdown](https://github.com/showdownjs/showdown). 

This addon adds **two unique** features on top of the `prism`  and `showdown` features. Firstly, if you provide a `data-filename` attribute, it will add this name to your code block including the file type. You’ve probably seen this before in the Ember Guides.  Secondly, line numbers are added to the code block and are also clickable which give you sharable links if you are using Fastboot or prember. 

[Chris](https://github.com/mansona) extracted the code from the guides, and with the help of [Nick Schot (@nickschot)](https://github.com/nickschot) managed to give you this cool addon with these awesome features.

Check out the [demo](https://ember-showdown-prism.netlify.app) and the [GitHub repo](https://github.com/empress/ember-showdown-prism) and add some cool code blocks to your blog or documentation!

---

## [RFC: {{id}} Helper 🆔](https://github.com/emberjs/rfcs/pull/659)

> [Steve Szczecina (@steveszc)](https://github.com/steveszc) proposed adding a new built-in template helper `{{id}}` for generating unique IDs. This helper will function as a replacement for the `elementId` used within classical Ember components, which is not available in glimmer components or route templates. Furthermore it powers common use cases such as:

Reviewer: ```suggestion
[Steve Szczecina (@steveszc)](https://github.com/steveszc) proposed adding a built-in [`{{id}}` helper](https://github.com/emberjs/rfcs/pull/659) for generating unique IDs. This helper serves as a replacement for `elementId`, which is available in classic components but not in Glimmer components or route templates. Furthermore, it powers common use cases such as:
```

> - Associating label and input elements using the label's for attribute and the input's id attribute.
> - Using WAI-ARIA attributes to improve accessibility (eg. aria-labelledby, aria-controls)
> - Integrating 3rd party libraries that attach themselves to DOM elements using DOM IDs (eg. maps, datepickers, jquery plugins, etc)

Reviewer: ```suggestion
- Associating label and input elements using the label's `for` attribute and the input's `id` attribute.
- Using WAI-ARIA attributes to improve accessibility (e.g. `aria-labelledby`, `aria-controls`)
- Integrating 3rd party libraries that attach themselves to DOM elements using DOM IDs (e.g. maps, calendars, jQuery plugins)
```

> The proposed `{{id}}` helper can be invoked with either no arguments or with the `for` argument. When no arguments are passed, it will return a new unique id string for every invocation. This invocation style would usually be paired with a `let` block.

Reviewer: ```suggestion
The proposed `{{id}}` helper can be invoked in 2 ways: without any argument or with a `for` argument.

When no arguments are passed, it will return a new unique ID string for every invocation. This invocation style would usually be paired with a `let` block.
```

```hbs
{{#let (id) as |emailId|}}
  <label for={{emailId}}>Email address</label>
  <input id={{emailId}} type="email" />
{{/let}}
```

<!--alex ignore savage-->

> The named argument `for` accepts any object, string, number, Element, or primitive, which will be treated as a stable reference for an id, allowing the helper to return the same id value for every invocation using the same for value.

Reviewer: ```suggestion
The named argument `for` accepts an object, string, number, Element, or primitive, and treats it as a stable reference for an ID. This allows the helper to return the same ID for a given input.
```

- Associating label and input elements using the label's `for` attribute and the input's `id` attribute.
- Using WAI-ARIA attributes to improve accessibility (e.g. `aria-labelledby`, `aria-controls`)
- Integrating 3rd party libraries that attach themselves to DOM elements using DOM IDs (e.g. maps, calendars, jQuery plugins)

The proposed `{{id}}` helper can be invoked in 2 ways: without any argument or with a `for` argument.

When no arguments are passed, it will return a new unique ID string for every invocation. This invocation style would usually be paired with a `let` block.

```hbs
{{#let (id) as |emailId|}}
  <label for={{emailId}}>Email address</label>
  <input id={{emailId}} type="email" />
{{/let}}
```

<!--alex ignore savage-->
The named argument `for` accepts an object, string, number, Element, or primitive, and treats it as a stable reference for an ID. This allows the helper to return the same ID for a given input.

```hbs
<label for="{{id for=this}}-email">Email address</label>
<input id="{{id for=this}}-email" type="email" />
```

> For more information on why we need the `{{id}}` helper [read the pre-RFC issue](https://github.com/emberjs/rfcs/issues/612) or [comment on the RFC](https://github.com/emberjs/rfcs/pull/659) on GitHub!

Reviewer: ```suggestion
For more information on why we need the `{{id}}` helper, please check out [RFC #659](https://github.com/emberjs/rfcs/pull/659). Be sure to provide your feedback by adding an emoji reaction or leaving a comment!
```
